openscenario/map_fragment/src/curve.cpp

```diff
@@ -63,33 +64,26 @@ try {
     return std::filesystem::path(node.get_parameter("output_directory").as_string());
   }();
 
-  auto lanelets = lanelet::Lanelets();
-
-  lanelets.push_back(makeLanelet(width, before_length, 0, resolution));
-  lanelets.push_back(makeLanelet(lanelets.back(), length, curvature, resolution));
-  lanelets.push_back(makeLanelet(lanelets.back(), after_length, 0, resolution));
-
-  for (auto i = 1; i < number_of_lanes; ++i) {
-    lanelets.push_back(makeLaneletRight(lanelets[lanelets.size() - 3], resolution));
-    lanelets.push_back(
-      makeLaneletRight(lanelets[lanelets.size() - 3], lanelets.back(), resolution));
-    lanelets.push_back(
-      makeLaneletRight(lanelets[lanelets.size() - 3], lanelets.back(), resolution));
-  }
-
-  lanelets.push_back(makeInvertedLaneletLeft(lanelets.back(), resolution));
-  lanelets.push_back(
-    makeInvertedLaneletLeft(lanelets[lanelets.size() - 3], lanelets.back(), resolution));
-  lanelets.push_back(
-    makeInvertedLaneletLeft(lanelets[lanelets.size() - 5], lanelets.back(), resolution));
-
-  for (auto i = 1; i < number_of_lanes; ++i) {
-    lanelets.push_back(makeLaneletLeft(lanelets[lanelets.size() - 3], resolution));
-    lanelets.push_back(makeLaneletLeft(lanelets[lanelets.size() - 3], lanelets.back(), resolution));
-    lanelets.push_back(makeLaneletLeft(lanelets[lanelets.size() - 3], lanelets.back(), resolution));
-  }
-
-  const auto map = lanelet::utils::createMap(lanelets);
+  const std::vector<ParametricCurve::ConstPointer> guide_curve_segments = {
+
+    // Straight segment before the turn
+    std::make_shared<Straight>(before_length),
+
+    // Turn segment (or straight if angle == 0)
+    angle == 0 ? std::static_pointer_cast<ParametricCurve>(  //
+                   std::make_shared<Straight>(length))
+               : std::static_pointer_cast<ParametricCurve>(
+                   std::make_shared<Arc>(length / std::abs(angle), angle)),
+
+    // Straight segment after the turn
+    std::make_shared<Straight>(after_length)};
```

Comment:
There was a rounding error because two integers were used for division instead of floating points. Because of that, all lanes were shifted to the left. This is inward in case of a left turn and resulted in reduced length.

Fixed in https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/8c533c1172354245e1a7e33ea6d1fba543f1c58a.

Length reported by `lane.cpp` is now 199.991. It is slightly lower than 200 because of discretization error and can be adjusted with the `resolution` parameter.

Reply:
I also applied additional changes so that `sample-curve.yaml` can be run:
- [I changed the unit of `angle` parameter back to degrees (I switched to radians at some point)](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/c3306657e16bf030c60015cd26b28d3896f50b7f);
- [I used a lower value of `route_length` in the launch file](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/1d85283878eb35f4cb7720824c6109b88c491cc4);
- [I removed `route_length` constraint in `find.cpp`](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/c9e7c0a9bd1482f543c4b5c4c01563513ebecc47) (for some reason it is not working with the generated map and can't find any candidates even when very short values are provided).

Reply:
The variable names `angle_in_degrees` and `angle_in_radians` are very good.

Reply:
@Goldob 
I ran `sample-curve.yaml` and it seems that the opposing lane is no longer generated (check base-branch's behavior).

Reply:
OK, I will add the opposing lanelets.

On a side note, the same parameter (`number_of_lanes`) has two different meanings in `lane.cpp` and `curve.cpp`. I the former case, it means the total number of lanes. In the latter, it means number of lanes per direction (so the total number is doubled). Perhaps we can try to make them consistent? Would the same behavior (double number of lanes) be preferable for both?

Reply:
@yamacir-kit 

[Done](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/994ba35dc8db12bb67d6413c585f8ced59fe8478). To implement this feature, I added a new field in `CrossSectionDescription` to specify how many lanes run in opposite direction (left-hand traffic is assumed). Behavior in `lane.cpp` for now I left unchanged.

[I also now needed to reduce `route_length` even more](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/76cb0c3082c130e33e54f086c02c8281d6f04dd6). This is because of the issue with `route_length_greater_than` constraint in `find.cpp`, I can't prevent the vehicle from driving on the inner side of the curve (200 is the length of road centerline, inner lanes will be shorter).

Reply:
@Goldob 

> On a side note, the same parameter (`number_of_lanes`) has two different meanings in `lane.cpp` and `curve.cpp`. In the former case, it means the total number of lanes. In the latter, it means number of lanes per direction (so the total number is doubled). Perhaps we can try to make them more consistent? Would the same behavior (double number of lanes) be preferable for both?

You are right, the meaning of `number_of_lanes` should be consistent. The reason for the lack of consistency is that `lane.cpp` and `curve.cpp` were developed by extending the functionality of `lane.cpp` and `curve.cpp` in turn, so `lane.cpp` does not have feedback of the functionality added from `curve.cpp`. I apologize for that.

For now, I believe the latter proposal should be selected. (That is not a task to be done in this pull request, but should be done at a later date.)

I think adding a parameter like `do_not_generate_opposing_lanes:=<boolean (defaultly false)>` in addition to `number_of_lanes` would make it easier to use. Here, I think we need to decide whether to default to the state with or without opposing lanes, based on the frequency of use.